File: 00002_meetup/java_compare/src/main/java/compare/java/Person.java
package compare.java;

public class Person {
    public final String firstName;
    public final String lastName;
    public final String petName;

    public Person(final String firstName, final String petName, final String lastName) {
        if(firstName == null) throw new IllegalArgumentException("first name is null");
        if(lastName == null) throw new IllegalArgumentException("last name is null");
        if(petName == null) throw new IllegalArgumentException("pet name is null");
        this.firstName = firstName;
        this.petName = petName;
        this.lastName = lastName;
    }

    public String toString() {
        return String.format("%s '%s' %s", firstName, petName, lastName);
    }

    @Override
    public boolean equals(final Object obj) {
        if(this == obj) {
            return true;
        } else if(obj instanceof Person) {
            final Person that = (Person) obj;
            //there is a possibility of a very subtle bug here.. and subtle bugs are hard to find! except if you did TDD but then ..
            return  this.firstName.equals(that.firstName) && this.petName.equals(that.petName)  && this.lastName.equals(that.lastName);
        } else {
            return false;
        }
    }

    //we need hashcode now...
}

interface Mortal {
    public String firstName();
}

File: 00002_meetup/java_compare/src/main/scala/Person.scala
package compare.scala

class Person(val firstName: String,  val lastName: String,  val petName: String)  extends Mortal {
  //demo how to use scala to reduce boilerplate.
  require(firstName != null, "first name's null")
  require(lastName != null, "last name's null")
  require(petName != null, "pet name's null")

  def someMethod() = "my string"

  //override is required
  override def toString() = "%s '%s' %s".format(firstName, petName, lastName)

  override def equals(obj: Any) = (obj.isInstanceOf[Person] && equalTo(obj.asInstanceOf[Person]))

  //I predicted it and I made it, find the error in the following line
  //private def equalTo(p: Person) = this.firstName == p.firstName && this.petName == p.lastName && this.lastName == p.lastName
  private def equalTo(p: Person) = this.firstName == p.firstName && this.petName == p.petName && this.lastName == p.lastName
}

object Person {
   val CONSTANT = "bla"
}

class Person2(val firstName: String,  val lastName: String,  val petName: String) extends Mortal {
  satisfies(Map(
    (firstName != null) -> "first name's null",
    (lastName != null) -> "last name's null",
    (petName != null) -> "pet name's null")
  )

  def satisfies(m: Map[Boolean, String]) = m.map(p => require(p._1, p._2))

  //override is required
  override def toString() = "%s '%s' %s".format(firstName, petName, lastName)
}

case class Individual(firstName: String,  lastName: String,  petName: String) extends Mortal

trait Mortal {
  def firstName: String
  def petName: String
  def lastName: String
}

trait MortalGreeter {
  def greet(m: Mortal): String
}

//aha traits can be used for implementations too! Mixins!
trait HelloGreeter extends MortalGreeter {
  def greet(m: Mortal) =
    "Hello mortal %s '%s' %s".format(m.firstName, m.petName, m.lastName)
}

//the following is a singleton
object HelloGreeter extends HelloGreeter

//mixin in action ->
class IndividualGreeter(override val firstName: String,  override val lastName: String, override val petName: String) extends Individual(firstName, lastName, petName) with HelloGreeter